kotlin/semlang-api/src/main/kotlin/net/semlang/api/language.kt

```diff
@@ -858,28 +855,60 @@ sealed class Expression {
+     * or conditionally may have an existing alias, and thus cannot be assigned to a variable if it's a reference type.
+     *
+     * Note that this category includes expressions that may be "definitely aliased", but there's no need to distinguish
+     * them from expressions that are "possibly aliased" (such as may result from an if-expression where one possible
+     * result is aliased and the other is not).
+     */
+    PossiblyAliased,
+    ;
+    companion object {
+        // I'd really rather this be named "for", but that's not an option
+        fun of(possiblyAliased: Boolean): AliasType {
+            if (possiblyAliased) {
+                return PossiblyAliased
+            } else {
+                return NotAliased
+            }
+        }
+    }
+}
 sealed class TypedExpression {
     abstract val type: Type
-    data class Variable(override val type: Type, val name: String): TypedExpression()
-    data class IfThen(override val type: Type, val condition: TypedExpression, val thenBlock: TypedBlock, val elseBlock: TypedBlock): TypedExpression()
-    data class NamedFunctionCall(override val type: Type, val functionRef: EntityRef, val resolvedFunctionRef: ResolvedEntityRef, val arguments: List<TypedExpression>, val chosenParameters: List<Type>, val originalChosenParameters: List<Type>): TypedExpression()
-    data class ExpressionFunctionCall(override val type: Type, val functionExpression: TypedExpression, val arguments: List<TypedExpression>, val chosenParameters: List<Type>, val originalChosenParameters: List<Type>): TypedExpression()
-    data class Literal(override val type: Type, val literal: String): TypedExpression()
-    data class ListLiteral(override val type: Type, val contents: List<TypedExpression>, val chosenParameter: Type): TypedExpression()
-    data class NamedFunctionBinding(override val type: Type, val functionRef: EntityRef, val resolvedFunctionRef: ResolvedEntityRef, val bindings: List<TypedExpression?>, val chosenParameters: List<Type?>, val originalChosenParameters: List<Type?>) : TypedExpression()
-    data class ExpressionFunctionBinding(override val type: Type, val functionExpression: TypedExpression, val bindings: List<TypedExpression?>, val chosenParameters: List<Type?>, val originalChosenParameters: List<Type?>) : TypedExpression()
-    data class Follow(override val type: Type, val structureExpression: TypedExpression, val name: String): TypedExpression()
-    data class InlineFunction(override val type: Type, val arguments: List<Argument>, val boundVars: List<Argument>, val returnType: Type, val block: TypedBlock): TypedExpression()
+    // TODO: Consider auto-setting some alias types; i.e. always aliased for Variable, never aliased for Literal/ListLiteral/maybe FunctionCalls
+    // (Non-obvious correction: Follows wouldn't be aliased if their structure expressions are unaliased, e.g. foo()->bar)
+    abstract val aliasType: AliasType
```

Comment:
We can consider making this getAliasType() or isPossiblyAliased() and give default implementations to some classes, but we probably want to wait and see how this ends up interacting with structs, functions, etc. first.